Rewrite ajax interface for the auth flow

danrschlosser opened the pull request:
Review: @eunicekokor @RaymondXu @natebrennand 

This is a code cleanup / refactoring PR.

Closes #113 
Closes #128 

Does the following:

- Change `json_response` to take only a dictionary to be JSON encoded
- Add `ajax_success` which returns a JSON shape containing some data:
    
    ```js
    {
        "status": "success",
        "data": <data:dict>
    }
    ```
- Add `ajax_error_message` which returns a JSON shape containing a message, the code, and optional extra data about the failure:
    
    ```js
    {
        "status": "error",
        "error": {
            "message": <error_message:string>,
            "code": <code:int>,
            "data": <error_data:dict>
        }
    }
    ```
- Refactor `auth.py` to use these interfaces on the server side
- Refactor `plus.js` to use these interfaces on the client side

See screenshots for examples of this code working correctly (as before)

![screen shot 2015-04-14 at 1 54 17 am](https://cloud.githubusercontent.com/assets/2433509/7131226/5111cc04-e24a-11e4-9166-d28c7800f9a8.png)
![screen shot 2015-04-14 at 1 26 32 am](https://cloud.githubusercontent.com/assets/2433509/7131225/50ee7772-e24a-11e4-90e7-bae1e222e404.png)

danrschlosser commented on app/static/js/admin/auth/plus.js:
```diff
@@ -39,25 +39,17 @@ var helper = (function() {
+                    window.location.href = result.data.redirect_url;
                 },
                 error: function(xhr, status, err) {
-                    if (xhr.status == 200) {
```
I think this can be removed... it was a bug that was happening because we weren't returning valid JSON, just a string, so jQuery thought it was a bad request....

danrschlosser commented:
@benlowkh any chance this code could be useful for you?

danrschlosser commented:
ping @natebrennand @jettca as I believe this is blocking for #193

jettca commented on app/lib/ajax.py:
```diff
@@ -0,0 +1,75 @@
+    :rtype: :class:`flask.Response`
+    """
+    if not error_data:
+        error_data = {}  # error_data should always be a dictionary
```
why not set error_date = {} by default, instead of None?

danrschlosser replied:
In Python, it's a hidden gotcha to use mutable default arguments in methods:

http://docs.python-guide.org/en/latest/writing/gotchas/

jettca commented on app/lib/ajax.py:
```diff
@@ -0,0 +1,75 @@
+    """Return a :class:`flask.Response` with a JSON encoded object ``data`` in
+    the body.
+
+    :param dict data: The data to be put in the response body.
```
this should in theory also work on lists, since json.dumps supports that. or at least it does with default set to json_util.default.

jettca commented on app/lib/ajax.py:
```diff
@@ -0,0 +1,75 @@
+    :returns: the response object
+    :rtype: :class:`flask.Response`
+    """
+    text = json.dumps(data)
```
add default=bson.json_util.default

danrschlosser commented:
Renamed everything from `ajax_*` to `json_*` and renamed `ajax.py` to `json.py`, in preparation for this library to be used in API responses like in #193

jettca commented:
ok lgtm
